feat(모달 띄우기): redux store에 modal 상태 생성 - #43

SkynI25 opened the pull request:
## 🚀 구현내용
- [x] redux store에 modal 상태 추가
  - [x] 테스트 코드 작성
- [x] component 폴더에 모달 컴포넌트 생성

redux를 이용해 모달 컴포넌트를 띄울 수 있도록 하였습니다. @Lavegaa @dididy 두 분의 도움에 감사드립니다! 👍 

## 🚀 궁금한 점
redux modal 패턴을 적용했을 때와 그렇지 않고 구현했을 때의 차이, 그리고 redux modal 패턴의 이점에 대해서 궁금하여 정리중에 있습니다. 관련 글들이 아직 찾고 있는 중인데..혹시 관련해서 알려주신다면 감사하겠습니다

## 🚀 어려운 점
- redux modal 패턴과 관련해 자료가 많지 않다는 느낌을 받았습니다.
  - [Modal을 Redux로 구현하기](https://velog.io/@altmshfkgudtjr/Redux%EB%A5%BC-%EC%9D%B4%EC%9A%A9%ED%95%9C-Modal-%EA%B5%AC%ED%98%84)
  - @Lavegaa 님과 함께 페어프로그래밍을 진행하면서 다음과 같은 단계로 redux를 활용해 모달을 띄웠습니다.
1. redux store에 모달들의 가시성 상태 key,value 형태로 추가(`questionListModal: false`)
2. Modal Container
    1. 각 modal view 컴포넌트 정보가 담긴 object가 존재
    2. Modal Container에 modal type을 인자로 전달함
    3. useSelector로 store에서 해당 모달의 상태값 가져옴
    4. 상태값에 따라서 조건부 렌더링함
3. Modal view 컴포넌트 작성하여 보여줌

위 링크를 참고하면서 진행 중인데 전체 예시 코드는 없는 상태라 하나씩 이해해가면서 구현하고 있어요 ㅎㅎ
- 그리고 redux-toolkit 을 함께 적용해가는 것도 약간 난관이 있지만 최대한 구현해보겠습니다 :)

Lavegaa committed:
feat: 글자로 이루어진 버튼 컴포넌트 제작 (#37)

Lavegaa committed:
feat: 질문 작성 기초 (#39)
Reactdnd연동
할일 - 클릭 이벤트 작성(애니메이션 등)

Lavegaa committed:
feat: 스크립트 작성 부분 완료 (#39)

Lavegaa committed:
feat: 오타 수정 (#39)

SkynI25 committed:
feat: 혼자연습하기 체크리스트 가이드 화면 제작- 기본화면 구현(#42)

SkynI25 committed:
feat(checkList): 혼자연습하기 체크리스트에 사용할 흰색 색상 추가 - #42

SkynI25 committed:
fix(lint 오류 수정): 세미콜론 추가

SkynI25 committed:
feat(checkList): 체크리스트 가이드 템플릿 제작 - #42

SkynI25 committed:
feat(checkList): 체크리스트 흰색 테두리 추가 - #42

SkynI25 committed:
feat(모달 띄우기): redux store에 modal 상태 생성 - #43

SkynI25 committed:
feat(모달 띄우기): redux store에 modal 상태 생성 - #43

Lavegaa committed:
feat: 주석 추가 (#39)

Lavegaa committed:
Merge pull request #38 from Yapp-17th/feat/#37

글자로 이루어진 버튼 컴포넌트 제작

Lavegaa commented:
https://github.com/Lavegaa/react-Blog/blob/master/blog-frontend/src/store/modules/base.js
좀 옛날에 만들어 봤던거긴 한데 한번 보시는거 추천드려요.
기본적으로 모달에 대해 dispatch날릴 때 
1. 모달을 show할것이다.
2. 어떤 모달 (type)을 보여줄 것이다.
3. 모달에 필요한 props가 있으면 같이 넘겨준다.

## 자세한 단계
1. redux state에서 visible을 관리합니다.
https://github.com/Lavegaa/react-Blog/blob/e1ab52da407ce7e54e0e6950edfc815218266054/blog-frontend/src/store/modules/base.js#L34
```js
//initial state
const initialState = Map({
    //모달의 가시성 상태
    modal: Map({
        remove: false,
        login: false
    }),
    loginModal: Map({
        password: '',
        error: false
    }),
    logged: false   //현재 로그인 상태
});
//reducer
export default handleActions({
    [SHOW_MODAL]: (state, action) => {
        const { payload: modalName } = action;
        return state.setIn(['modal', modalName], true);
    },
    [HIDE_MODAL]: (state, action) => {
        const { payload: modalName } = action;
        return state.setIn(['modal', modalName], false);
    },
```
2. 모달container에서 visible을 받아 모달 컴포넌트에 전달해줍니다.
여기에서 container / view 구조를 썼기때문에 container가 있는거지 필수는 아닙니다.
https://github.com/Lavegaa/react-Blog/blob/e1ab52da407ce7e54e0e6950edfc815218266054/blog-frontend/src/containers/modal/LoginModalContainer.js#L46
```js
    render() {
        const {
            handleCancel, handleChange, handleKeyPress, handleLogin
        } = this;
        const { visible, error, password } = this.props;

        return (
            <LoginModal
                onLogin={handleLogin} onCancel={handleCancel}
                onChange={handleChange} onKeypress={handleKeyPress}
                visible={visible} error={error} password={password}
            />
        );
    }
}

export default connect(
    (state) => ({
        visible: state.base.getIn(['modal', 'login']),
        password: state.base.getIn(['loginModal', 'password']),
        error: state.base.getIn(['loginModal', 'error'])
    }),
    (dispatch) => ({
        BaseActions: bindActionCreators(baseActions, dispatch)
    })
)(LoginModalContainer);
```
3.  visible일 때만 모달을 return 한다.
https://github.com/Lavegaa/react-Blog/blob/e1ab52da407ce7e54e0e6950edfc815218266054/blog-frontend/src/components/modal/ModalWrapper/ModalWrapper.js#L34
```js
   render() {
        const { children, visible } = this.props;
        const { animate } = this.state;
        //visible과 animate값이 둘다 false일 때만
        //null을 리턴
        if(!visible && !animate) return null;

        //상태에 따라 애니메이션 설정
        const animation = animate && (visible ? 'enter' : 'leave');

        return (
            <div>
                <div className={cx('gray-background')}/>
                <div className={cx('modal-wrapper')}>
                    <div className={cx('modal', animation)}>
                        {children}
                    </div>
                </div>
            </div>
        );
    }
}
```

Lavegaa committed:
fix: lock파일 추가(#39)

voidmatcha committed:
env(actions): 배포 브랜치 url에 따라 따로 볼 수 있는 GitHub Actions 설정 추가 (#49)
- Branch 이름 얻기
  - echo "::set-env name=variable_name::content" 이거는 Deprecate 되어서 사용하면 안됨
  - echo "{name}={value}" >> $GITHUB_ENV 를 사용해야 함
  - name은 ${{ env.name }} 형식으로 사용할 수 있음
  - 브랜치 이름에 '#'이 포함되어 있으므로 이를 제거해야 함
    - tr을 이용하려 했으나 공백으로 치환이 불가능하다는 문제가 생김
    - cut으로 대체하려 했으나 구분자를 기준으로 분리된 것 중 하나를 선택하는 형태라 실패
    - 결국 sed를 활용해 #을 공백으로 치환함
- PR 혹은 master에 merge시 브랜치 이름 폴더에 빌드되도록 destination_dir 추가
- commit_message url에 브랜치 이름을 포함하여 커밋된 브랜치에 따른 URL 포함하도록 함

https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable

https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/

Commit:
Merge pull request #50 from Yapp-17th/env/#49

Storybook 배포 브랜치별로 따로 볼 수 있도록 GitHub Actions 설정 추가

voidmatcha committed:
fix(actions): master branch 경로는 merge 후 배포시 접속 안되는 문제 해결 (#49)
- 조건을 걸어서 master branch 인 경우 root 에 빌드하도록 수정함

Commit:
Merge pull request #52 from Yapp-17th/env/#49

master branch 경로는 merge 후 배포시 접속 안되는 문제 해결

voidmatcha committed:
fix(actions): PR 브랜치별 스토리북 배포를 위한 keep_files 활성화
- if를 사용하도록 수정해서 keep_files 옵션을 켜야 빌드 폴더들이 유지됨

voidmatcha committed:
Merge branch 'env/#49'

SkynI25 commented:
@Lavegaa 감사합니다!! 한번 참고해서 작성해보겠습니다 :)

Lavegaa commented:
넵 redux자체가 옛날 스타일이여서 redux-tools를 활용하면 더 깔끔한 구조를 가져갈 수 있을 것 같아요. 
immutable관련된 것 도 지원해주니 확실히 깔끔해질 것 같습니다.

voidmatcha committed:
fix(eslint): rules , 구분자 오타 수정 (#55)
Merge하는 과정에서 발생한 문제라고 생각됨

Commit:
Merge pull request #56 from Yapp-17th/fix/#55

eslint rules의 , 구분자 중 하나가 사라짐

voidmatcha committed:
feat: 스크롤바 디자인 적용 (#54)
- ::-webkit-scrollbar, ::-webkit-scrollbar-track, ::-webkit-scrollbar-thumb 을 사용해서 디자인 적용
- storybook에서도 적용되도록 .storybook/preview-head.html에 css 추가

Commit:
Merge pull request #57 from Yapp-17th/feat/#54

스크롤바 디자인 적용

Lavegaa committed:
feat: mock파일 작성. 주석제거. (#39)

Lavegaa committed:
Merge pull request #48 from Yapp-17th/feat/#39

질문 리스트 컴포넌트 제작

SkynI25 committed:
feat(비디오 스타일링) - 비디오 태그 스타일링, 예제 코드 적용 #42

SkynI25 committed:
feat(CheckList): 체크리스트 가이드 부분 추가 - #42

SkynI25 committed:
feat(modal 구현): ModalWrapper 배치 - #43

SkynI25 committed:
feat(modal 구현): ModalWrapper 배치 - #43

SkynI25 committed:
feat(modal 구현): ModalWrapper 생성 - #43

SkynI25 committed:
feat(modal 구현): ModalWrapper 생성 - #43

SkynI25 committed:
feat(modal 구현): modal redux 내용 수정(element, props, style 추가) - #43

SkynI25 committed:
feat(modal 구현): modal redux 내용 수정(element, props, style 추가) - #43

SkynI25 committed:
refactor(리팩토링) - 불필요한 코드 제거 - #43

SkynI25 committed:
feat(모달 제작): 각 모달에 대한 상수값들을 별도의 constant 파일로 분리

SkynI25 committed:
feat(모달 제작): 모달 컨테이너 제작(모달 타입 전달받아서 해당하는 모달 조건부 렌더링함) - #43

SkynI25 committed:
feat(모달 제작): test 용 모달 컴포넌트 제작

SkynI25 committed:
feat(모달 제작): 메인화면에서 버튼 클릭하면 모달 보여지도록 함 - #43

SkynI25 committed:
feat(모달 제작): 각 모달의 상태(boolean) 만들고 모달을 보여주고 지워주는 thunk 함수 생성함 - #43

SkynI25 committed:
Merge branch 'feat/#43' of https://github.com/Yapp-17th/Web_2_Client into feat/#43

SkynI25 committed:
refactor(리팩토링): 불필요한 코드 제거 - #43

voidmatcha reviewed (approved):
Modal 관련된 부분 확인하였습니다. redux hooks 잘 사용하셨네요. 승인합니다!

Lavegaa commented on src/pages/AloneQuestionCheckList/AloneQuestionCheckList.js:
```diff
@@ -0,0 +1,343 @@
+    	text-align:center;
+    	overflow:hidden;
+    	white-space:nowrap;
+      	text-overflow:ellipsis;
```
여기 밀린 부분이 있습니다~

SkynI25 replied:
확인해보겠습니다 !

Lavegaa commented on src/store/Modal/modal.js:
```diff
@@ -0,0 +1,35 @@
+  },
+});
+
+export const { displayModal, removeModal } = modalReducer.actions;
```
혹시 아래 thunk패턴?으로 export하신거 테스트 해보셨나요??
제 파트에서 작성할 때 해당 패턴으로 작성하다 안되서 기존 각 컴포넌트에서 dispatch하는걸로 작성했었는데
thunk패턴으로 잘 돌아가면 방법좀 부턱드립니다.. 분명 같이봤는데 저는 안되더라고요ㅋㅋㅋ
그리고 만약 thunk패턴으로 dispatch들을 export해준다면 actions는 따로 export할 필요가 없을 것 같아요.

SkynI25 replied:
> 혹시 아래 thunk패턴?으로 export하신거 테스트 해보셨나요??
제 파트에서 작성할 때 해당 패턴으로 작성하다 안되서 기존 각 컴포넌트에서 dispatch하는걸로 작성했었는데
thunk패턴으로 잘 돌아가면 방법좀 부턱드립니다

저는 export 한 thunk함수를 컴포넌트에서 dispatch 하면서 사용했습니다! thunk 함수를 컴포넌트에서 어떻게 사용했는지를 질문하신것으로 이해를 했는데, 맞나요? 아래는 thunk 함수를 적용한 컴포넌트 코드의 일부입니다.
```jsx
return (
    <>
      <Modal modalName={MODALS.QUESTIONLIST_SAVE_MODAL} />
      <button type="button" onClick={() => handleCreate()}>컨퍼런스 생성</button>
      <button onClick={() => dispatch(showModal(MODALS.QUESTIONLIST_SAVE_MODAL))} type="button">Modal open</button>
      <button onClick={() => dispatch(hideModal(MODALS.QUESTIONLIST_SAVE_MODAL))} type="button">Modal close</button>
    </>
  );
```


> 만약 thunk패턴으로 dispatch들을 export해준다면 actions는 따로 export할 필요가 없을 것 같아요.

말씀하신대로 현재 코드에선 actions 함수를 다른 곳에서 사용하지 않기 때문에 export 할 필요 없는거 같네요, 수정하겠습니다!

Lavegaa replied:
```js
return (
    <>
      <Modal modalName={MODALS.QUESTIONLIST_SAVE_MODAL} />
      <button type="button" onClick={() => handleCreate()}>컨퍼런스 생성</button>
      <button onClick={() => dispatch(showModal(MODALS.QUESTIONLIST_SAVE_MODAL))} type="button">Modal open</button>
      <button onClick={() => dispatch(hideModal(MODALS.QUESTIONLIST_SAVE_MODAL))} type="button">Modal close</button>
    </>
  );
```
위와 같은 패턴으로 사용한다면 결국 dispatch로 함수를 보내는건가요..? 기존 action을 보내는것과 어떤 차이가 있는지 궁금합니다.
```js
export const showModal = (modalName) => (dispatch) => {
  dispatch(displayModal({ modalName }));
};
```
위와같이 미리 dispatch한 것을 export한다면 따로 dispatch에 담지 않고 바로  import해서 사용이 가능 할 줄 알았는데 dispatch에 담아서 함수를 보내주네요. 어렵네요 정말 허허..

SkynI25 replied:
> 위와 같은 패턴으로 사용한다면 결국 dispatch로 함수를 보내는건가요..? 기존 action을 보내는것과 어떤 차이가 있는지 궁금합니다.

이해하신게 정확합니다. 기존 action 을 보내는 것과 구체적인 차이는 없는 것으로 보이는데요, thunk 를 사용했을 때의 이점을 정리해봐야 할 거 같아요.
> 위와같이 미리 dispatch한 것을 export한다면 따로 dispatch에 담지 않고 바로 import해서 사용이 가능 할 줄 알았는데 dispatch에 담아서 함수를 보내주네요. 

저도 이렇게 생각했는데 dispatch를 해줘야 하더군요 (여기서 삽질을 조금..) 말씀하신대로 일반적으로 actions 함수를 export 해서 사용했을 때와 큰 차이가 없어보이네요. thunk 를 사용하는 이유, 이점에 대해서 한번 정리해서 공유해보겠습니다! :)

Lavegaa replied:
감사합니다. 고생 많으섰어요!

SkynI25 replied:
https://www.softkraft.co/how-to-setup-redux-with-redux-toolkit/
영어라는 압박이 쵸큼 있지만 가볍게 읽어보시면 좋을 거 같아요 ㅎㅎ

SkynI25 committed:
refactor: 코드리뷰 피드백 반영 - #43

SkynI25 committed:
refactor: eslint 적용하여 문법에러 수정

SkynI25 committed:
refactor: 불필요한 공백 제거 - #43
